apisix/plugins/file-logger.lua

```diff
@@ -0,0 +1,212 @@
+end
+
+
+local function reopen()
+    local args = ngx.req.get_uri_args()
+    if args["reopen"] then
+        local target_pointer, err = io_open('logs/file_pointer', 'r')
+        if not target_pointer then
+            core.log.error("failed to open file, error info: " .. err)
+        else
+            local pointer = target_pointer:read('*n')
+            target_pointer:close()
+            core.log.warn('pointer: ', pointer)
+            C.close(pointer)
+
+            local reset_pointer = io_open('logs/file_pointer', 'w+')
+            reset_pointer:write(-1)
+            reset_pointer:close()
+        end
+    end
+end
+
+
+local function write_file_data(conf, log_message)
+    local msg = core.json.encode(log_message) .. "\n"
+    local fd = file_descriptors[conf.path]
+    local file = io_open(conf.path, 'r')
+    local file_pointer = io_open('logs/file_pointer', 'r')
+
+    if file_pointer then
+        local read_pointer = file_pointer:read('*n')
+        core.log.warn('read_pointer: ', read_pointer)
+        file_pointer:close()
+        if read_pointer < 0 then
+            C.close(fd)
+            fd = nil
+            file_descriptors[conf.path] = nil
+        end
+    else
+        fd = file_descriptors[conf.path]
+    end
+
+    if not file then
+        file = io_open(conf.path, 'a+')
+        file:close()
+        if fd then
+            C.close(fd)
+            fd = nil
+            file_descriptors[conf.path] = nil
+        end
+    else
+        file:close()
+    end
+
+
+    if not fd then
+        fd = C.open(conf.path, oflags, mode)
+        core.log.warn('fd: ', fd)
+        local write_pointer = io_open('logs/file_pointer', 'w+')
+        write_pointer:write(fd)
+        write_pointer:close()
+
+        if fd < 0 then
+            local err = ffi.errno()
+            core.log.error("failed to open file: " .. conf.path .. ", error info: " .. err)
+
+        else
+            file_descriptors[conf.path] = fd
+        end
+    end
+
+    C.write(fd, msg, #msg)
+end
+
+
+function _M.body_filter(conf, ctx)
+    log_util.collect_body(conf, ctx)
+end
+
+
+function _M.log(conf, ctx)
+    local metadata = plugin.plugin_metadata(plugin_name)
+    local entry
+
+    if metadata and metadata.value.log_format
+        and core.table.nkeys(metadata.value.log_format) > 0
+    then
+        entry = log_util.get_custom_format_log(ctx, metadata.value.log_format)
+    else
+        entry = log_util.get_full_log(ngx, conf)
+    end
+
+    if conf.custom_fields_by_lua
+        and core.table.nkeys(conf.custom_fields_by_lua) > 0
+    then
+        local set_log_fields_value = entry
+        for key, expression in pairs(conf.custom_fields_by_lua) do
+            set_log_fields_value[key] = load_string(expression)()
+        end
+        write_file_data(conf, set_log_fields_value)
+    else
+        write_file_data(conf, entry)
+    end
+end
+
+
+function _M.control_api()
+    return {
+        {
+            methods = {"GET"},
```

Comment:
What about using `PUT`?
We already use it in the reload: https://github.com/apache/apisix/blob/master/docs/en/latest/architecture-design/plugin.md#hot-reload